MG-1965 - Process Event Logs

rodneyosodo opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

<!--

Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of the issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

For Work In Progress Pull Requests, please use the Draft PR feature, see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments.

- Provide tests for your changes.
- Use descriptive commit messages.
- Comment your code where appropriate.
- Squash your commits
- Update any related documentation.
-->

# What type of PR is this?

This is a feature PR because it adds eventlogs service to store history of events

<!--This represents the type of PR you are submitting.

For example:
This is a bug fix because it fixes the following issue: absmach/magistrala#1234
This is a feature because it adds the following functionality: ...
This is a refactor because it changes the following functionality: ...
This is a documentation update because it updates the following documentation: ...
This is a dependency update because it updates the following dependencies: ...
This is an optimization because it improves the following functionality: ...
-->

## What does this do?

Adds event logs service
<!--
Please provide a brief description of what this PR is intended to do.
Include List any changes that modify/break current functionality.
-->

## Which issue(s) does this PR fix/relate to?

<!--
For pull requests that relate or close an issue, please include them below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "Resolves absmach/magistrala#1234" would connect the current pull request to issue 1234.  And when we merge the pull request, Github will automatically close the issue.
-->

- Related Issue #
- Resolves absmach/magistrala-ui-new#164 absmach/magistrala#1957

## Have you included tests for your changes?

Yes, I have included tests for my changes.

<!--If you have not included tests, please explain why.
For example:
Yes, I have included tests for my changes.
No, I have not included tests because I do not know how to.
-->

## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

<!--If you have not included documentation, please explain why.
For example:
Yes, I have updated the documentation for the new feature.
No, I have not updated the documentation because I do not know how to.
-->

### Notes

<!--Please provide any additional information you feel is important.-->

rodneyosodo committed:
Add event logs service

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

ianmuchyri commented on api/openapi/event-logs.yml:
```diff
@@ -0,0 +1,278 @@
+          - domain
+          - thing
+          - platform
+          - thing
```
```suggestion

```

rodneyosodo replied:
Yes thing is a valid type

ianmuchyri replied:
It's repeated

rodneyosodo committed:
feat: add with_payload parameter

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

ianmuchyri commented on eventlogs/postgres/events.go:
```diff
@@ -0,0 +1,150 @@
+	if page.WithPayload {
+		sq += ", payload"
+	}
+	q := fmt.Sprintf("SELECT %s FROM events %s ORDER BY occurred_at LIMIT :limit OFFSET :offset;", sq, query)
```
we should display events from the latest. Order by `DESC`

rodneyosodo replied:
Added direction as a query parameter with the default being descending.

rodneyosodo committed:
feat: add direction parameter

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
use slog for logging as mglogger was deprecated

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
fix after rebase

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

dborovcanin commented on Makefile:
```diff
@@ -5,7 +5,7 @@ MG_DOCKER_IMAGE_NAME_PREFIX ?= magistrala
 SERVICES = auth users things http coap ws lora influxdb-writer influxdb-reader mongodb-writer \
 	mongodb-reader cassandra-writer cassandra-reader postgres-writer postgres-reader timescale-writer timescale-reader cli \
-	bootstrap opcua twins mqtt provision certs smtp-notifier smpp-notifier invitations
+	bootstrap opcua twins mqtt provision certs smtp-notifier smpp-notifier invitations event-logs
```
Rename to `activities` or similar alternative.

dborovcanin commented on cmd/event-logs/main.go:
```diff
@@ -0,0 +1,174 @@
+
+	logger.Info("Subscribed to Event Store")
+
+	if err := eventlogs.Start(ctx, svcName, subscriber, repo); err != nil {
```
Add event processing to service, rather than using `repo` directly. Also, add logging to it.

dborovcanin commented on eventlogs/events.go:
```diff
@@ -0,0 +1,73 @@
+//go:generate mockery --name Service --output=./mocks --filename service.go --quiet --note "Copyright (c) Abstract Machines"
+type Service interface {
+	// ReadAll retrieves all events from the database with the given page.
+	ReadAll(ctx context.Context, token string, page Page) (EventsPage, error)
```
Add `Save` and use service instead of repo for consuming.

dborovcanin commented on eventlogs/api/transport.go:
```diff
@@ -0,0 +1,103 @@
+
+	mux := chi.NewRouter()
+
+	mux.Get("/events/{id}/{type}", otelhttp.NewHandler(kithttp.NewServer(
```
ID and Type should be optional query params.

dborovcanin reviewed (changes requested):
Not a message consumer, but this is still a consumer, so let's move it there.

rodneyosodo committed:
rename service to activitylog

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

WashingtonKK commented on cli/activitylog.go:
```diff
@@ -0,0 +1,50 @@
+	},
+}
+
+// NewActivitiesCmd returns invitations command.
```
```suggestion
// NewActivitiesCmd returns activities command.
```

WashingtonKK commented on mqtt/events/events.go:
Update also https://github.com/rodneyosodo/magistrala/blob/f34d4a24f611e6b99da06d37a0ced37cde4e06fc/internal/groups/events/events.go#L61-L69 .

WashingtonKK replied:
The `unassignGroupEvent` lacks an id, how will you store it in the activities log?

rodneyosodo replied:
reverting back

rodneyosodo committed:
refactor(activitylog): remove id from activity

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
refactor(activitiylog): change format to activity

- Added attributes and metadata as seperate values of activity. Metadata is normal metadata for an entity while attributes are the other entities from the event for example id, updated_at e.t.c
- add entityType which can either be empty or user, group, thing or channel
- add query method which generates SQL query depending on the entity type for example is the entity is user we retrieve entities that have id as the entity id and have operation that is prefixed with user. OR events that have user_id as an attribute correcsponding to that entity
- update emission and consumption of events. Emit native types for example string slice for tags rather than formating it and also metadata as it is rather than marshalling it first into bytes

This contains breaking changes on how events are consumed

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

arvindh123 commented on internal/groups/events/events.go:
```diff
@@ -4,14 +4,13 @@
 )
 
-const (
+var (
```
Should we use `var` here ?

rodneyosodo replied:
We are using `var`

arvindh123 commented on activitylog/postgres/init.go:
```diff
@@ -0,0 +1,32 @@
+						attributes	JSONB NOT NULL,
+						metadata	JSONB,
+						UNIQUE(operation, occurred_at, attributes)
+					)`,
```
We can have indexing here like `CREATE INDEX ON activities (operation, occurred_at);`

rodneyosodo replied:
I have created the following index for testing purpose. 

```SQL
CREATE INDEX idx_activities_group_filter_sort ON activities(operation, (attributes->>'id'), (attributes->>'group_id'), occurred_at DESC);
```

And found out that creating an index actually hurts our total DB performance mostly on counting total. What do you think?

This is the trace report for groups given the command

```bash
curl -X GET -H "Authorization: Bearer $USER_TOKEN" "http://localhost:9021/activities?id=$GROUP_ID&entity_type=group&with_attributes=true"
```

![image](https://github.com/absmach/magistrala/assets/28790446/7578a515-e01b-44bc-867e-4c54373aebaf)

For a different query which has not been index ie

```bash
curl -X GET -H "Authorization: Bearer $USER_TOKEN" "http://localhost:9021/activities?id=$THING_ID&entity_type=thing&with_attributes=true"
```

This is the trace report

![image](https://github.com/absmach/magistrala/assets/28790446/d71d6f57-3ff7-4b2f-9778-a638829573b7)

rodneyosodo replied:
Actually after indexing we have faster query times

arvindh123 commented on activitylog/consumer.go:
```diff
@@ -0,0 +1,82 @@
+	subCfg := events.SubscriberConfig{
+		Consumer: consumer,
+		Stream:   store.StreamAllEvents,
+		Handler:  Handle(service),
```
Can we have something like `service.EventHandler`? 

And its definition will be like

```go
func (svc *service)EventHandle(ctx context.Context, event events.Event) error {
	data, err := event.Encode()
	if err != nil {
		return err
	}

	operation, ok := data["operation"].(string)
	if !ok {
		return errors.New("missing operation")
	}
	delete(data, "operation")

	if operation == "" {
		return errors.New("missing operation")
	}

	occurredAt, ok := data["occurred_at"].(float64)
	if !ok {
		return ErrMissingOccurredAt
	}
	delete(data, "occurred_at")

	if occurredAt == 0 {
		return ErrMissingOccurredAt
	}

	metadata, ok := data["metadata"].(map[string]interface{})
	if !ok {
		metadata = make(map[string]interface{})
	}
	delete(data, "metadata")

	if len(data) == 0 {
		return errors.New("missing attributes")
	}

	activity := Activity{
		Operation:  operation,
		OccurredAt: time.Unix(0, int64(occurredAt)),
		Attributes: data,
		Metadata:   metadata,
	}

	return svc.repository.Save(ctx, activity)
}

```

rodneyosodo replied:
We would lose more information when logging and tracing since event has not yet been processed to an activity

arvindh123 commented on activitylog/activity.go:
```diff
@@ -0,0 +1,165 @@
+type Service interface {
+	// Save saves the activity to the database.
+	Save(ctx context.Context, activity Activity) error
+
```
As suggested before,  here can we have something like `EventHandler` https://github.com/absmach/magistrala/pull/2057#discussion_r1589139829

arvindh123 reviewed (changes requested):
User of another domain able to see the events all of all domain and non domain events. 

With the below request `user_2` whois in `domain_2` and not in `domain_1` able to see events of `domain_1`

```bash
curl --location 'localhost:9021/activities?with_attributes=true' \
--header 'Authorization: Bearer <TOKEN>'
```

We should have domain as sperate column in table, so that we can add indexing and also helps in filter based on domain.

And we need additional column to identify it is domain releated event or non domain event. 

We can reuse the domain column with const value non-domain-event to identify the non domain event 


Non domain event should able to access by SuperAdmins 

Domain related login events should able to access ony by Domain Admins or Editors 

Entity editor should able to see the entity activities and viewer and members should not able to see the activities of domain.

All the above consideration will lead to new challenge , List all events for particular domain will becomes a challenge, I suggestion something like , for now we can list events of particular entity id and  to list all events on particular domain, we can do it later after having local index caching of spicedb policy or after some other mechanism until we find.

rodneyosodo committed:
fix: failing tests due to change in events structure

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(activity-log): add db indexes

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

rodneyosodo committed:
feat(activitylog): add schemathesis

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>

WashingtonKK commented on internal/apiutil/errors.go:
```diff
@@ -173,4 +173,13 @@ var (
+	ErrMissingEntityType = errors.New("missing entity type")
+
+	// ErrInvalidTime indicates invalid time i.e not unix time.
+	ErrInvalidTime = errors.New("invalid time use unix time")
```
Reword this to something like: "invalid time format" or something that is clearer to understand.

rodneyosodo committed:
refactor(activitylog): Refactor authorization

A normal user can be able to list invitations but they will get only the operation and time it occurred unless you are the platform admin where you can view anything

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

WashingtonKK reviewed (commented):
A user in `domainA` can see activities of a separate `domainB` even though they are not part of the domain. Is this intended?
Can we have auth such that a user can only see the activities in their domain?
@rodneyosodo what do you think?

arvindh123 commented on activitylog/service.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	return svc.repository.RetrieveAll(ctx, page)
+}
```
@rodneyosodo Do you think we should all activities if user is not super admin ? 
If yes, why we should all activites for non super admin, How it will be useful from them .

rodneyosodo replied:
What is being retrieved doesn't contain sensitive information since with omit attributes and metadata we only retrieve the operation and when it occurred

arvindh123 replied:
I'm just thinking, How it will be usefull for user. It might usefull for case like how in `ecommerce` site they will show popup with hidden sensitive infromation of recently happened purchases.
But how it might be useful much in IoT case ?. 

So can we make it configurable in env with by default `false`? 

So if application want to show all events infromation, they can enable it by env variable

rodneyosodo committed:
fix(activitylog): enforce `/{type}/{id}` and remove `/activities`

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

arvindh123 commented on activitylog/api/endpoint_test.go:
```diff
@@ -0,0 +1,282 @@
+
+	for _, c := range cases {
+		t.Run(c.desc, func(t *testing.T) {
+			repoCall := svc.On("RetrieveAll", mock.Anything, c.token, mock.Anything).Return(activitylog.ActivitiesPage{}, c.svcErr)
```
`svcCall` might be a better name

rodneyosodo committed:
fix: rename repocall to svcall

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

arvindh123 commented on activitylog/activity.go:
```diff
@@ -0,0 +1,162 @@
+	case ThingEntity:
+		return "((operation LIKE 'thing.%' AND attributes->>'id' = :entity_id) OR (attributes->>'thing_id' = :entity_id))"
+	case ChannelEntity:
+		return "((operation LIKE 'channel.%' AND attributes->>'id' = :entity_id) OR (attributes->>'group_id' = :entity_id))"
```
@rodneyosodo 
I could not list `channels`create  event. 
I think this should be `group` or we should change in all events to `channel`

rodneyosodo committed:
fix: abstract user with channel as a valid type

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

arvindh123 commented on activitylog/service.go:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	return nil
+}
```
This can be done this way, In this way you could not view user activities without domain token 

To view users there is no mandatory for domain token, both login token and domain token should work. 

But here with login token you could not view the user activities. It is because login token doesn't have subject, So the authorization fails. 


So I suggest something like below,
But drawback of this method is it requires two auth calls. 
This problem exits all over the project, So there is ticket to reslove this two auth call issue. 

```suggestion
func (svc *service) authorize(ctx context.Context, token, entityID, entityType string) error {
	res, err := svc.auth.Identify(ctx, &magistrala.IdentityReq{Token: token})
	if err != nil {
		return err
	}

	permission := auth.ViewPermission
	objectType := entityType
	object := entityID
	subject := res.GetId()
	// If the entity is a user, we need to check if the user is an admin
	if entityType == auth.UserType {
		permission = auth.AdminPermission
		objectType = auth.PlatformType
		object = auth.MagistralaObject
		subject = res.GetUserId()
	}

	req := &magistrala.AuthorizeReq{
		SubjectType: auth.UserType,
		SubjectKind: auth.UsersKind,
		Subject:     subject,
		Permission:  permission,
		ObjectType:  objectType,
		Object:      object,
	}

	authRes, err := svc.auth.Authorize(ctx, req)
	if err != nil {
		return errors.Wrap(svcerr.ErrAuthorization, err)
	}
	if !authRes.GetAuthorized() {
		return svcerr.ErrAuthorization
	}

	return nil
}
```

rodneyosodo replied:
Why should someone view activities without a domain login, this is because you can't view an entity without it too?

arvindh123 replied:
Users are not not owned by domain, they are owned by Platform
So `SuperAdmin` should able to see the users activity logs without login to domain. 
Like how we can able to list domain and list users without login to domain.

arvindh123 commented on cli/activitylog.go:
```diff
@@ -0,0 +1,50 @@
+
+var cmdActivity = cobra.Command{
+	Use:   "get <entity_id> <entity_type> <user_auth_token>",
+	Short: "Get activities",
```
```suggestion
	Use:   "get <entity_type> <entity_id> <user_auth_token>",
```

rodneyosodo committed:
fix: start with entitytype then ID

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix: change authorization for users

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(bootstrap): use shared event read and readSlice funcs

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on activitylog/service.go:
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Abstract Machines
```
Add a doc.go explaining the purpose of the service.

dborovcanin commented on activitylog/postgres/init.go:
```diff
@@ -0,0 +1,36 @@
+				Up: []string{
+					`CREATE TABLE IF NOT EXISTS activities (
+						id 			UUID NOT NULL DEFAULT gen_random_uuid() PRIMARY KEY,
+						operation	VARCHAR NOT NULL CHECK (operation <> ''),
```
No need for this check, since we don't do that in the rest of the codebase. We will perform this check during validation.

dborovcanin reviewed (changes requested):
Please rename the service, Journal seems like a good name (Log is suboptimal, and Activities and Events could be easily misinterpreted).

rodneyosodo committed:
refactor(journal): Rename from activitlog to journal

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
fix(journal): Remove empty check for operation's field

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
docs(journal Add doc.go): Add go documentation for the journal service, events and middleware package in journal service

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

dborovcanin commented on journal/service.go:
```diff
@@ -0,0 +1,77 @@
+
+type service struct {
+	auth       magistrala.AuthServiceClient
+	repository Repository
```
Let's use a UUID provider like in other services.

rodneyosodo committed:
fix(journal): Remove id generation from db

Move ID generation from database to service layer

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
